src/panel/direction/DirectionPanel.jsx

```diff
@@ -65,17 +67,32 @@ export default class DirectionPanel extends React.Component {
     this.dragPointHandler = listen('change_direction_point', this.changeDirectionPoint);
     this.setPointHandler = listen('set_direction_point', this.setDirectionPoint);
 
-    if (!this.props.origin && isMobileDevice()) {
-      // If authorized, set origin to current position on mobile
-      const origin = new NavigatorGeolocalisationPoi();
-      try {
-        await origin.geolocate({ displayErrorModal: false });
-        this.setState(
-          { origin, originInputText: origin.name },
-          this.update
-        );
-      } catch (e) {
-        // ignore possible error
+
+    // on mobile, when no origin is specified, try auto-geoloc
+    if (isMobileDevice() && !this.state.origin && !this.props.origin) {
+      const geolocationPermission = await getGeolocationPermission();
+      let modalAccepted = false;
+
+      // on an empty from, if the user's position permission hasn't been asked yet, show modal
+      if (!this.state.destination && !this.props.destination
```

Comment:
ok, it's easier to understand now. The code might be a bit complicated with nested if and try/catch, but it's ok